README.md
<div align="center">
    <a href="https://github.com/nvzqz/cargo-emit">
        <img src="https://raw.githubusercontent.com/nvzqz/cargo-emit/assets/logo.svg?sanitize=true"
             width="300"
             height="300">
    </a>
    <h1>Cargo Emit</h1>
    <a href="https://crates.io/crates/cargo-emit">
        <img src="https://img.shields.io/crates/v/cargo-emit.svg" alt="Crates.io">
        <img src="https://img.shields.io/crates/d/cargo-emit.svg" alt="Downloads">
    </a>
    <a href="https://travis-ci.com/nvzqz/cargo-emit">
        <img src="https://travis-ci.com/nvzqz/cargo-emit.svg?branch=master" alt="Build Status">
    </a>
    <img src="https://img.shields.io/badge/rustc-^1.31.0-blue.svg" alt="rustc ^1.31.0">
    <br>
    <a href="https://www.patreon.com/nvzqz">
        <img src="https://c5.patreon.com/external/logo/become_a_patron_button.png" alt="Become a Patron!" height="35">
    </a>
    <a href="https://www.paypal.me/nvzqz">
        <img src="https://buymecoffee.intm.org/img/button-paypal-white.png" alt="Buy me a coffee" height="35">
    </a>
    <br><br>
</div>

Talk to Cargo easily at build time, brought to you by [Nikolai Vazquez].

This library provides:

- Convenience macros for communicating with Cargo during the [`build.rs`]
  phrase. Cargo listens to certain [build script outputs] that dictate how
  it should behave.

- An accessible location for seeing what script build outputs are available
  to emit.

- Protection against typos that can be made when printing these formatted
  outputs directly yourself. Mistyping macro names will result in a compile
  failure.

[Nikolai Vazquez]:      https://twitter.com/NikolaiVazquez
[`build.rs`]:           https://doc.rust-lang.org/cargo/reference/build-scripts.html
[build script outputs]: https://doc.rust-lang.org/cargo/reference/build-scripts.html#outputs-of-the-build-script

## Usage

This crate exposes the following macros:

| Macro                                  | Output                                |
|----------------------------------------|---------------------------------------|
| [`pair!($key, $value)`]                | `cargo:$key=$value`                   |
| [`rerun_if_changed!($path)`]           | `cargo:rerun-if-changed=$path`        |
| [`rerun_if_env_changed!($key)`]        | `cargo:rerun-if-env-changed=$key`     |
| [`rustc_cdylib_link_arg!($flag)`]      | `cargo:rustc-cdylib-link-arg=$flag`   |
| [`rustc_cfg!($feature)`]               | `cargo:rustc-cfg=$feature`            |
| [`rustc_env!($key, $value)`]           | `cargo:rustc-env=$key=$value`         |
| [`rustc_flags!($flags)`]               | `cargo:rustc-flags=$flags`            |
| [`rustc_link_arg!($arg)`]              | `cargo:rustc-link-arg=$arg`           |
| [`rustc_link_arg_bin!($bin => $arg)`]  | `cargo:rustc-link-arg-bin=$bin=$arg`  |
| [`rustc_link_arg_bins!($arg)`]         | `cargo:rustc-link-arg-bins=$arg`      |
| [`rustc_link_lib!($name => $kind)`]    | `cargo:rustc-link-lib=$kind=$name`    |
| [`rustc_link_search!($path => $kind)`] | `cargo:rustc-link-search=$kind=$path` |
| [`warning!($message)`]                 | `cargo:warning=$message`              |

[`pair!($key, $value)`]:                  https://docs.rs/cargo-emit/latest/cargo_emit/macro.pair.html
[`rerun_if_changed!($path)`]:      https://docs.rs/cargo-emit/latest/cargo_emit/macro.rerun_if_changed.html
[`rerun_if_env_changed!($key)`]:  https://docs.rs/cargo-emit/latest/cargo_emit/macro.rerun_if_env_changed.html
[`rustc_cdylib_link_arg!($flag)`]: https://docs.rs/cargo-emit/latest/cargo_emit/macro.rustc_cdylib_link_arg.html
[`rustc_cfg!($feature)`]:             https://docs.rs/cargo-emit/latest/cargo_emit/macro.rustc_cfg.html
[`rustc_env!($key, $value)`]:             https://docs.rs/cargo-emit/latest/cargo_emit/macro.rustc_env.html
[`rustc_flags!($flags)`]:           https://docs.rs/cargo-emit/latest/cargo_emit/macro.rustc_flags.html
[`rustc_link_arg!($arg)`]:        https://docs.rs/cargo-emit/latest/cargo_emit/macro.rustc_link_arg.html
[`rustc_link_arg_bin!($bin => $arg)`]:    https://docs.rs/cargo-emit/latest/cargo_emit/macro.rustc_link_arg_bin.html
[`rustc_link_arg_bins!($arg)`]:   https://docs.rs/cargo-emit/latest/cargo_emit/macro.rustc_link_arg_bins.html
[`rustc_link_lib!($name => $kind)`]:        https://docs.rs/cargo-emit/latest/cargo_emit/macro.rustc_link_lib.html
[`rustc_link_search!($path => $kind)`]:     https://docs.rs/cargo-emit/latest/cargo_emit/macro.rustc_link_search.html
[`warning!($message)`]:               https://docs.rs/cargo-emit/latest/cargo_emit/macro.warning.html

## License

This project is released under either:

- [MIT License](https://github.com/nvzqz/cargo-emit/blob/master/LICENSE-MIT)

- [Apache License (Version 2.0)](https://github.com/nvzqz/cargo-emit/blob/master/LICENSE-APACHE)


CHANGELOG.md
# Changelog [![Crates.io][crate-badge]][crate]

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog] and this project adheres to
[Semantic Versioning].

## [Unreleased]

## [0.2.1] - 2021-09-01

### Fixed

- Added `#[allow(clippy::explicit_write)]` to prevent false-positives.
- Added `examples/macros.rs` file and corresponding `cargo clippy --example "macros" -- -D warnings` CI job to prevent future clippy issues.

## [0.2.0] - 2021-08-31

### Fixed

- Fixed recursion in `rustc_link_search!(…)`.
- Removed redundant `$x:literal` macro patterns.

### Added

- [`rustc_link_arg!`](https://docs.rs/cargo-emit/0.2.0/cargo_emit/macro.rustc_link_arg.html)
- [`rustc_link_arg_bin!`](https://docs.rs/cargo-emit/0.2.0/cargo_emit/macro.rustc_link_arg_bin.html)
- [`rustc_link_arg_bins!`](https://docs.rs/cargo-emit/0.2.0/cargo_emit/macro.rustc_link_arg_bins.html)
- 40 snapshot tests (using [insta](https://crates.io/crates/insta))
- Optional `to:` parameter for writing into a `std::fmt::Write` or `io::fmt::Write`.

# Removed

- `#![no_std]`, since the project never worked on no_std to begin with.

## [0.1.1] - 2019-12-15

### Fixed

- Links to docs for specific macros.

## 0.1.0 - 2019-12-15

### Added

- [`pair!`](https://docs.rs/cargo-emit/0.1.0/cargo_emit/macro.pair.html)
- [`rerun_if_changed!`](https://docs.rs/cargo-emit/0.1.0/cargo_emit/macro.rerun_if_changed.html)
- [`rerun_if_env_changed!`](https://docs.rs/cargo-emit/0.1.0/cargo_emit/macro.rerun_if_env_changed.html)
- [`rustc_cdylib_link_arg!`](https://docs.rs/cargo-emit/0.1.0/cargo_emit/macro.rustc_cdylib_link_arg.html)
- [`rustc_cfg!`](https://docs.rs/cargo-emit/0.1.0/cargo_emit/macro.rustc_cfg.html)
- [`rustc_env!`](https://docs.rs/cargo-emit/0.1.0/cargo_emit/macro.rustc_env.html)
- [`rustc_flags!`](https://docs.rs/cargo-emit/0.1.0/cargo_emit/macro.rustc_flags.html)
- [`rustc_link_lib!`](https://docs.rs/cargo-emit/0.1.0/cargo_emit/macro.rustc_link_lib.html)
- [`rustc_link_search!`](https://docs.rs/cargo-emit/0.1.0/cargo_emit/macro.rustc_link_search.html)
- [`warning!`](https://docs.rs/cargo-emit/0.1.0/cargo_emit/macro.warning.html)

[crate]:       https://crates.io/crates/cargo-emit
[crate-badge]: https://img.shields.io/crates/v/cargo-emit.svg

[Keep a Changelog]:    http://keepachangelog.com/en/1.0.0/
[Semantic Versioning]: http://semver.org/spec/v2.0.0.html

[Unreleased]: https://github.com/nvzqz/static-assertions-rs/compare/v0.1.1...HEAD
[0.1.1]: https://github.com/nvzqz/static-assertions-rs/compare/v0.1.0...v0.1.1


src/lib.rs
//! <div align="center">
//!     <a href="https://github.com/nvzqz/cargo-emit">
//!         <img src="https://raw.githubusercontent.com/nvzqz/cargo-emit/assets/logo.svg?sanitize=true"
//!              alt="Cargo Emit Logo"
//!              width="300"
//!              height="300">
//!     </a>
//!     <br><br>
//!     <a href="https://crates.io/crates/cargo-emit">
//!         <img src="https://img.shields.io/crates/d/cargo-emit.svg"
//!              alt="Downloads">
//!     </a>
//!     <a href="https://travis-ci.com/nvzqz/cargo-emit">
//!         <img src="https://travis-ci.com/nvzqz/cargo-emit.svg?branch=master"
//!              alt="Build Status">
//!     </a>
//!     <img src="https://img.shields.io/badge/rustc-^1.37.0-blue.svg"
//!          alt="rustc ^1.37.0">
//!     <br><br>
//! </div>
//!
//! Talk to Cargo easily at build time, brought to you by [Nikolai Vazquez].
//!
//! This library provides:
//!
//! - Convenience macros for communicating with Cargo during the [`build.rs`]
//!   phrase. Cargo listens to certain [build script outputs] that dictate how
//!   it should behave.
//!
//! - An accessible location for seeing what script build outputs are available
//!   to emit.
//!
//! - Protection against typos that can be made when printing these formatted
//!   outputs directly yourself. Mistyping macro names will result in a compile
//!   failure.
//!
//! # Usage
//!
//! This crate is available [on crates.io][crate] and can be used by adding the
//! following to your project's [`Cargo.toml`]:
//!
//! ```toml
//! [build-dependencies]
//! cargo-emit = "0.1"
//! ```
//!
//! and something like this to your [`build.rs`]:
//!
//! ```
//! # let should_warn = true;
//! if should_warn {
//!     cargo_emit::warning!("(C-3PO voice) We're doomed");
//! }
//! ```
//!
//! **Note:** This library is meant to be used with [Rust 2018 edition][2018],
//! so that `cargo_emit::` can be used to prefix macro calls.
//!
//! # Compatibility
//!
//! This crate is compatible with Rust 1.31+ in order to use the
//! `$crate::macro!` feature introduced in [Rust 2018][2018].
//!
//! # Examples
//!
//! Very thorough examples are provided in the docs for
//! [each individual macro](#macros).
//!
//! # Donate
//!
//! This project is made freely available (as in free beer), but unfortunately
//! not all beer is free! So, if you would like to buy me a beer (or coffee or
//! *more*), then consider supporting my work that's benefited your project
//! and thousands of others.
//!
//! <a href="https://www.patreon.com/nvzqz">
//!     <img src="https://c5.patreon.com/external/logo/become_a_patron_button.png" alt="Become a Patron!" height="35">
//! </a>
//! <a href="https://www.paypal.me/nvzqz">
//!     <img src="https://buymecoffee.intm.org/img/button-paypal-white.png" alt="Buy me a coffee" height="35">
//! </a>
//!
//! [Nikolai Vazquez]: https://twitter.com/NikolaiVazquez
//! [build script outputs]: https://doc.rust-lang.org/cargo/reference/build-scripts.html#outputs-of-the-build-script
//! [2018]: https://blog.rust-lang.org/2018/12/06/Rust-1.31-and-rust-2018.html#rust-2018
//! [crate]: https://crates.io/crates/cargo-emit
//! [`Cargo.toml`]: https://doc.rust-lang.org/cargo/reference/manifest.html
//! [`build.rs`]: https://doc.rust-lang.org/cargo/reference/build-scripts.html

#![doc(html_root_url = "https://docs.rs/cargo-emit/0.1.1")]
#![doc(
    html_logo_url = "https://raw.githubusercontent.com/nvzqz/cargo-emit/assets/logo.svg?sanitize=true"
)]
#![deny(missing_docs)]

/// `cargo:$key=$value`
mod pair;
/// `cargo:rerun-if-changed=$path`
mod rerun_if_changed;
/// `cargo:rerun-if-env-changed=$key`
mod rerun_if_env_changed;
/// `cargo:rustc-cdylib-link-arg=$flag`
mod rustc_cdylib_link_arg;
/// `cargo:rustc-cfg=$feature`
mod rustc_cfg;
/// `cargo:rustc-env=$key=$value`
mod rustc_env;
/// `cargo:rustc-flags=$flags`
mod rustc_flags;
/// `cargo:rustc-link-arg=$arg`
mod rustc_link_arg;
/// `cargo:rustc-link-arg-bin=$bin=$arg`
mod rustc_link_arg_bin;
/// `cargo:rustc-link-arg-bins=$arg`
mod rustc_link_arg_bins;
/// `cargo:rustc-link-lib=$kind=$name`
mod rustc_link_lib;
/// `cargo:rustc-link-search=$kind=$path`
mod rustc_link_search;
/// `cargo:warning=$message`
mod warning;

#[cfg(test)]
fn capture_output<F>(f: F) -> String
where
    F: FnOnce(&mut String),
{
    let mut output = String::new();
    f(&mut output);
    output
}
